proposals/0075-precompile-for-secp256r1-sigverify.md

```diff
@@ -0,0 +1,298 @@
+## New Terminology
+
+None
+
+## Detailed Design
+
+The precompile's purpose is to verify signatures using ECDSA-256.
+(denoted in [RFC6460](https://www.ietf.org/rfc/rfc6460.txt) as
+ECDSA using the NIST P-256 curve and the SHA-256 hashing algorithm)
+
+Apart from the RFC mandated implementation the precompile must additionally take
+an opinionated stance on signature malleability.
+
+### Signature Malleability
+
+Due to X axis symmetry along the elliptic curve, for any ECDSA signature
+`(r, s)`, there also exists a valid signature `(r, n - s)`, where `n` is the
+order of the curve. This introduces "s malleability", allowing an attacker
+to produce an alternative version of `s` without invalidating the signature.
+
+The pitfalls of this in authentication systems can be particularly perilous,
+opening up certain implementations to signature replay attacks over the same
+message by simply flipping the `s` value over the curve.
+
+As the primary goal of the `secp256r1` program is secure signature validation
+for authentication purposes, the precompile must mitigate these attacks
+by enforcing the usage of `lowS` values, in which `s <= n/2`.
+
+As such, the program must immediately fail upon the detection of any
+signature that includes a `highS` value. This prevents any accidental
+succeptibility to signature malleability attacks.
+
+Note: The existing `secp256k1` precompile makes no attempt attempt to mitigate
+s malleability, as doing so would go against its primary goal of achieving
+`ecrecover` parity with EVM.
+
+### Implementation
+
+### Program
+
+ID: `Secp256r1SigVerify1111111111111111111111111`
+
+The program instruction must be composed of the following struct:
+
+In Pseudocode:
+
+```
+struct Secp256r1SigVerifyInstruction {
+    count: uint8 LE,                          // Number of signatures to check
+    padding: uint8 LE,                        // Single byte padding
+    signatureOffsets: Array<Secp256r1SignatureOffsets>, // Array of structs
+}
+
+struct Secp256r1SignatureOffsets {
+    signature_offset: uint16 LE,              // Offset to signature
+    signature_instruction_index: uint8 LE,   // Instruction index to signature
+    public_key_offset: uint16 LE,             // Offset to public key
+    public_key_instruction_index: uint8 LE,  // Instruction index to  public key
+    message_data_offset: uint16 LE,           // Offset to start of message data
+    message_data_size: uint16 LE,             // Size of message data
+    message_instruction_index: uint8 LE,     // Instruction index to message
+}
+```
+
+Multiple signatures can be verified. If any of the signatures fail to verify,
+an error must be returned.
+
+If the instruction data is empty, the program must return an error.
+
+If `count == 0` and the length of the instruction data > 1, the program must
+return an error.
+
+If `count == 0` and the length of the instruction data is not at least
+`count * Secp256r1SignatureOffsets + 1`, the program must return an error.
```

Comment:
This should be 2 to include the padding byte, right? You should also make it clear that `Array` does not contain any padding between elements or any length prefixes.

```suggestion
If `count > 0` and the length of the instruction data is not at least
`count * sizeof(Secp256r1SignatureOffsets) + 2`, the program must return an error.
```

Reply:
absolutely correct, fixing 👍 

will also note the lack of length prefix and padding between elements, thanks for the suggestion 👍